pkg/scheduler/admission/podnodeconstraints/admission.go

```diff
@@ -0,0 +1,201 @@
+	}
+	return err
+}
+
+// extract the PodSpec from the pod templates for each object we care about
+func (o *podNodeConstraints) getPodSpec(attr admission.Attributes) (kapi.PodSpec, error) {
+	switch r := attr.GetObject().(type) {
+	case *kapi.Pod:
+		return r.Spec, nil
+	case *kapi.ReplicationController:
+		return r.Spec.Template.Spec, nil
+	case *extensions.Deployment:
+		return r.Spec.Template.Spec, nil
+	case *extensions.ReplicaSet:
+		return r.Spec.Template.Spec, nil
+	case *extensions.Job:
+		return r.Spec.Template.Spec, nil
+	case *deployapi.DeploymentConfig:
+		return r.Spec.Template.Spec, nil
+	}
+	return kapi.PodSpec{}, kapierrors.NewInternalError(fmt.Errorf("No PodSpec available for supplied admission attribute"))
+}
+
+// validate PodSpec if NodeName or NodeSelector are specified
+func (o *podNodeConstraints) admitPodSpec(attr admission.Attributes, ps kapi.PodSpec) error {
+	matchingLabels := []string{}
+	// nodeSelector blacklist filter
+	if len(ps.NodeSelector) > 0 {
+		for nodeSelectorLabel := range ps.NodeSelector {
+			for _, blacklistLabel := range o.config.NodeSelectorLabelBlacklist {
+				if blacklistLabel == nodeSelectorLabel {
+					matchingLabels = append(matchingLabels, blacklistLabel)
+				}
+			}
+		}
+	}
+	// nodeName constraint
+	if len(ps.NodeName) > 0 || len(matchingLabels) > 0 {
+		allow, err := o.checkPodsBindAccess(attr)
+		if err != nil {
+			return err
+		}
+		if !allow {
+			switch {
+			case len(ps.NodeName) > 0 && len(matchingLabels) == 0:
+				return admission.NewForbidden(attr, fmt.Errorf("node selection by nodeName is prohibited by policy for your role"))
+			case len(ps.NodeName) == 0 && len(matchingLabels) > 0:
+				return admission.NewForbidden(attr, fmt.Errorf("node selection by label(s) %v is prohibited by policy for your role", matchingLabels))
+			case len(ps.NodeName) > 0 && len(matchingLabels) > 0:
+				return admission.NewForbidden(attr, fmt.Errorf("node selection by nodeName and label(s) %v is prohibited by policy for your role", matchingLabels))
+			}
+		}
+	}
+	return nil
+}
+
+// func (o *podNodeConstraints) SetOpenshiftClient(c client.Interface) {
+// 	o.client = c
+// }
+
+func (o *podNodeConstraints) SetAuthorizer(a authorizer.Authorizer) {
+	o.authorizer = a
+}
+
+func (o *podNodeConstraints) Validate() error {
+	// if o.client == nil {
+	// 	return fmt.Errorf("PodNodeConstraints needs an Openshift client")
+	// }
+	if o.authorizer == nil {
+		return fmt.Errorf("PodNodeConstraints needs an Openshift Authorizer")
+	}
+	return nil
+}
+
+// build LocalSubjectAccessReview struct to validate role via checkAccess
+func (o *podNodeConstraints) checkPodsBindAccess(attr admission.Attributes) (bool, error) {
+	ctx := kapi.WithUser(kapi.WithNamespace(kapi.NewContext(), attr.GetNamespace()), attr.GetUserInfo())
+
+	authzAttr := authorizer.DefaultAuthorizationAttributes{
+		Verb:         "create",
+		Resource:     "pods/binding",
+		APIGroup:     kapi.GroupName,
+		ResourceName: attr.GetName(),
```

Comment:
This bit of information only makes sense when you're operating on a Pod.  Otherwise the name is a non-sequitur.